app/scripts/modules/core/src/search/widgets/Search.tsx

```diff
@@ -0,0 +1,365 @@
+    const result: ITag[] = [];
+    Object.keys(queryParams).forEach((key: string) => {
+      const filterType: IFilterType = SearchFilterTypeRegistry.getFilterType(key);
+      if (filterType) {
+        result.push({
+          modifier: filterType.modifier,
+          text: queryParams[key] as string
+        });
+      }
+    });
+
+    return result;
+  }
+
+  private setFocusState(isFocused: boolean): void {
+    this.setState({
+      isFocused: isFocused
+    });
+  }
+
+  private isTagAlreadyPresent(tag: ITag): boolean {
+    return this.state.tags.some((t: ITag) => t.modifier === tag.modifier && t.text === tag.text);
+  }
+
+  private hasModifier(input: string): boolean {
+    const index = input.indexOf(':');
+    return this.modifiers.has(input.substring(0, index).toLocaleLowerCase()) && (index !== input.length);
+  }
+
+  private getModifier(input: string): string {
+    return input.substring(0, input.indexOf(':')).toLocaleLowerCase();
+  }
+
+  private getSearchText(input: string): string {
+
+    let result = input;
+    if (this.hasModifier(input)) {
+      result = input.substring(input.indexOf(':') + 1, input.length);
+    }
+
+    return result;
+  }
+
+  private buildTagFromInputString(): ITag {
+
+    let result: ITag = null;
+    const value = this.inputElement.value.trim();
+    if (value) {
+      const text = this.getSearchText(value).trim();
+      let modifier = this.state.activeFilter.modifier;
+      if (text) {
+        if (this.hasModifier(value)) {
+          modifier = this.getModifier(value);
+        }
+
+        result = { modifier, text }
+      }
+    }
+
+    return result;
+  }
+
+  private getActiveFilterIndex(): number {
+    return this.filterTypes.findIndex((type: IFilterType) => type === this.state.activeFilter);
+  }
+
+  private getActiveFilterText(filterType: IFilterType): string {
+
+    let result: string;
+    const text = this.inputElement.value.trim();
+    if (this.hasModifier(text)) {
+      const modifier = this.getModifier(text);
+      const regex = new RegExp(modifier, 'i');
+      result = `${text.replace(regex, filterType.modifier)}`;
+    } else {
+      result = `${filterType.modifier}:${text}`;
+    }
+
+    return result;
+  }
+
+  private handleTagListUpdate(elements: HTMLElement[]): void {
+    this.tagElements = elements;
+  }
+
+  private refCallback(element: HTMLInputElement) {
+    this.inputElement = element;
+  }
+
+  private handleBlur(): void {
+    this.setFocusState(false);
+    this.setState({
+      isOpen: false
+    });
+  }
+
+  private handleChange(): void {
+    const value = this.inputElement.value.trim();
+    if (this.hasModifier(value)) {
+      const modifier = this.getModifier(value);
+      this.setState({
+        activeFilter: this.filterTypes.find((type: IFilterType) => type.modifier === modifier),
+        isOpen: true
+      });
+    } else if (value) {
+      this.setState({
+        isOpen: true
+      });
+    } else {
+      this.setState({
```

Comment:
Not a big deal, just a pattern that I like is to have one `this.setState()` call in the function so it's easier to grok that there really is only one setState happening.